text/3389-manifest-lint.md

```diff
@@ -0,0 +1,398 @@
+- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
+- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new `lints` table would be added to configure lints:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+and `cargo` would pass these along as flags to `rustc` and `clippy`.
+
+This would work with
+[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
+```toml
+[lints]
+workspace = true
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+## Documentation Updates
+
+## The `lints` section
+
+*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*
+
+Override the default level of lints from different tools by assigning them to a new level in a
+table, for example:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+
+This is short-hand for:
+```toml
+[lints.rust]
+unsafe = { level = "forbid", priority = 0 }
+```
+
+`level` corresponds to the lint levels in `rustc`:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+`priority` is a signed integer that controls which lints or lint groups override other lint groups:
+- lower (particularly negative) numbers have lower priority, being overridden
+  by higher numbers, and show up first on the command-line to tools like
+  `rustc`
+
+To know which table under `[lints]` a particular lint belongs under, it is the part before `::` in the lint
+name.  If there isn't a `::`, then the tool is `rust`.  For example a warning
+about `unsafe` would be `lints.rust.unsafe` but a lint about
+`clippy::enum_glob_use` would be `lints.clippy.enum_glob_use`.
+
+## The `lints` table
+
+*as a new [`[workspace]` entry](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section)*
+
+The `workspace.lints` table is where you define lint configuration to be inherited by members of a workspace.
+
+Specifying a workspace lint configuration is similar to package lints.
+
+Example:
+
+```toml
+# [PROJECT_DIR]/Cargo.toml
+[workspace]
+members = ["crates/*"]
+
+[workspace.lints.rust]
+unsafe = "forbid"
+```
+
+```toml
+# [PROJECT_DIR]/crates/bar/Cargo.toml
+[package]
+name = "bar"
+version = "0.1.0"
+
+[lints]
+workspace = true
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When parsing a manifest, cargo will resolve workspace inheritance for
+`lints.workspace = true` as it does with basic fields, when `workspace` is
+present, no other fields are allowed to be present.  This precludes having the
+package override the workspace on a lint-by-lint basis.
+
+cargo will contain a mapping of tool to underlying command (e.g. `rust` to
+`rustc`, `clippy` to `rustc` when clippy is the driver, `rustdoc` to
```

Comment:
Yes, Clippy calls `cargo fix`, but with Clippy lints registered in the linked `rustc` driver, so that they get run and fixed by `cargo fix`.

I wouldn't say that cargo is responsible for applying the lints. Cargo can just pass all the specified lints to the `driver` that is used. `rustc` will ignore all tool lints that are not registered. Clippy ignores `-W`/... flags altogether and just forwards those to the `rustc` driver.

I'm not sure anymore if I understand the problem statement here. I think `cargo` doesn't have to filter lints by the tool it belongs to, but can just always specify all lints.

Reply:
> I'm not sure anymore if I understand the problem statement here. I think cargo doesn't have to filter lints by the tool it belongs to, but can just always pass all specified lints.

cargo fingerpints all build flags to know when to rebuild.  One shortcoming of `RUSTFLAGS` is that all of them are included in fingerpinting even if they are not relevant.  Can we use the extra context that cargo has with the `lints` table to reduce rebuilds when lints not relevant to the current driver are changed.

> I wouldn't say that cargo is responsible for applying the lints. Cargo can just pass all the specified lints to the driver that is used. rustc will ignore all tool lints that are not registered. Clippy ignores -W/... flags altogether and just forwards those to the rustc driver.

What is was referring to is that cargo will be responsible for reading `[lints]` and passing them to the driver.

I will say that though cargo does identify the driver and key some UX off of that (to specialize the message on how to fix),  that doesn't mean that it is reliable enough to make behavior choices off of.  I'm also unsure how registering of custom linters would affect this.